### DISCUSSION TITLE
How to access layout exposed methods from pages?

### DISCUSSION THREAD

**Discussion by sibevin** [nuxt/nuxt#17350] [Questions]
I have a layout called `dashboard` with an exposed method `showAlert()` as below:

_layouts/dashboard.vue_
```vue
<script setup lang="ts">
const showAlert = () => {
  console.log("showAlert() in dashboard");
};
defineExpose({ showAlert });
</script>
``` 

And try to call it from a page called `account` which is using the `dashboard` layout:

_pages/account.vue_
```vue
<script setup lang="ts">
onMounted(() => {
  // try to trigger showAlert() from here ....
});
</script>
<template>
  <NuxtLayout name="dashboard">
  </NuxtLayout>
</template>
```
I try to use `ref` to access the layout component as below:

_pages/account.vue_
```vue
<script setup lang="ts">
const dashboardLayout = ref();
onMounted(() => {
  console.log(dashboardLayout.value); // The layout can be accessed, but it is not the `dashboard` component indeed ...
  console.log(dashboardLayout.value.showAlert); // undefined 
  dashboardLayout.value.showAlert(); // TypeError: dashboardLayout.value.showAlert is not a function
});
</script>
<template>
  <NuxtLayout ref="dashboardLayout" name="dashboard">
  </NuxtLayout>
</template>
```
`ref` works on normal components, so it seems that the layout component has different mechanism. Are there any ways to access layout exposed methods from pages? Thanks.

nuxt version: `nuxt@3.0.0-rc.4`

**Comment by valcosmos** [comment]
I achieved a similar effect, it does work, hope it can helps you.

If this is not what u want, u can try pinia.

`layouts/dashboard.vue`

```vue
<script setup lang="ts">
  const showAlert = () => {
    console.log('showAlert() in dashboard')
  }
  
  const emit = defineEmits(['show-alert'])
  
  emit('show-alert', showAlert)
</script>
```

`pages/account.vue`

```vue
<script setup lang="ts">
  const showAlert = (fn) => {
    fn()
  }
</script>

<template>
 <NuxtLayout @show-alert="showAlert" name="demo">
  </NuxtLayout>
</template>
```

**Reply by sibevin** [reply]
Thank you for the reply. I tried this solution but have some problems in my real use case:
1. The `showAlert()` method need to access variables / components in the `dashboard` component. When `showAlert()` is triggered by `show-alert` event in `account`, these variables / components cannot be accessed anymore and cause the `undefined` errors. 
2. The `show-alert` event is triggered when `dashboard` component is loaded. In the real use case, we need to trigger it when we want to show some alerts.

IMO, it is a little weird because the `show-alert` event is triggered by `dashboard`, not by `account`, and it is actually opposite to which we want to achieve here.

**Reply by valcosmos** [reply]
Hello, I guess u may need to do component communication. Probably using Pinia would be a good choice.

**Reply by sibevin** [reply]
Yeah, we can use state management such as `Pinia` or nuxt-native `useState` to do component communication indeed, but it seems overkill. The exposed method works well on normal components and is also more straightforward in this use case. If it is not supported by layouts, we may consider using the `dashboard` component directly instead of using it as a layout, but we will lose some benefits like switching layout dynamically as well.

**Reply by valcosmos** [reply]
Hello, I found another way to do it that seems like a little weird🤣, but it does seem to work.

![image](https://user-images.githubusercontent.com/53572196/177232961-e923ebca-35c3-4fe4-84bf-87c4245df5f1.png)

`pages/account.vue`

```vue
<script setup lang="ts">
const alertNum = ref(0)

onMounted(() => {
  console.log('on mounted')

  alertNum.value++
})
</script>

<template>
  <NuxtLayout :alert="alertNum" name="dashboard"></NuxtLayout>
</template>

```

`layouts/dashboard.vue`

```vue
<script lang="ts" setup>
const showAlert = () => {
  console.log('showAlert() in dashboard')
}

const props = defineProps({
  alert: {
    type: Number,
    default: 0
  }
})

watch(
  () => props.alert,
  () => {
    showAlert()
  }
)
</script>
```

**Reply by sibevin** [reply]
I have tried this workaround and it actually works!! (It also means that the dynamic prop works on layouts.) I think this is the best approach we can get so far. Although this workaround has an obvious drawback that the number of props and watchers increases drastically if more exposed methods are required, especially when there are many parameters in each of them.

**Reply by valcosmos** [reply]
Hope it helped you, best of luck.😊
